### PR TITLE
test-consensus: refactor PBft ChainState tests

### PR DESCRIPTION
**Pull request by nfrisby**
This is a follow-up to PR #1317. That added EBBs to the PBFT chain state and to the corresponding property tests. This PR is a more thorough update of those tests: it's clearer code and improves coverage some.

It's a pretty big diff. Several of the properties are essentially the same. The generator is also essentially the same, but it's more principled and better documented by the Haddock on the data type being generated. In particular, it almost exclusively relies on only two splitting functions (from the new `Test.Util.Split` module) instead of various takes, drops, filters, reverses, etc.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 1]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
+--
+-- We first generate three segments of 'Inputs'.
+--
+-- >                  B
+-- >              ---------
+-- >       A     /
+-- >   ---------<
+-- >             \
+-- >              ---------
+-- >                  C
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Can you draw `C` longer than `B`, as it will contain at least `n` more blocks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 2]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
+
+    -- Pick number of signed slots
+    --
+    -- We don't try to be too clever here; we need to test the various edge
+    -- cases near genesis more carefully, but also want to test where we are
+    -- well past genesis
+    numSignersAB <- oneof [
+        choose (0, k + 1)
+      , choose (0, n + 1)
+      , choose (0, k + n + 1)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Why the `+ 1`s? Is it to cover the other side of the interesting bound as well?

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 3]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
+
+    -- Generate all the inputs
+
+    (inA, inB, inC) <- do
+        inA <- generateInputs genKey numSignersA (LatestInput Nothing)
+
+        -- Segment B must not begin with the same slot that A ended with,
+        -- because rewinds can't split a slot and our tests focus on the result
+        -- of a rewind that can be undone by re-applying B.
+        let linpA' = LatestInput $ case unLatestInput $ lastInput inA of
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: `lastInputInA'` instead of `linpA'` might be clearer.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 4]
```diff
@@ -427,61 +572,87 @@ appendInputs ::
...
+safeLast :: [a] -> Maybe a
+safeLast
+ = go
+  where
+    go = \case
+      []     -> Nothing
+      [x]    -> Just x
+      _ : xs -> go xs
+
+-- | <https://en.wikipedia.org/wiki/Monus>
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Ah, I did not know this, interesting!

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 5]
```diff
@@ -427,61 +572,87 @@ appendInputs ::
...
-dropWhileNotAfterInputs mSlot =
-  Inputs . dropWhile ((<= mSlot) . At . slotInput) . unInputs
+lastInput :: Inputs c -> LatestInput c
+lastInput = LatestInput . safeLast . unInputs
 
-takeLastInputs :: Word64 -> Inputs c -> Inputs c
-takeLastInputs n = Inputs . takeLast (fromIntegral n) . unInputs
+-- | The slot of the latest block
+slotLatestInput :: LatestInput c -> WithOrigin SlotNo
+slotLatestInput (LatestInput mbInp) = maybe Origin (At . slotInput) mbInp
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Would `Ouroboros.Network.Point.withOriginFromMaybe` make this clearer/

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 6]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
+
+        -- the state before PR #1307 didn't retain as many signatures and
+        -- didn't track EBBs at all
+        oldABb = fromInputs paramK paramN anchor
+                   (signatureInputs inps)
+                   mempty
+          where
+            inps = snd $ splitAtSigner (numSignersA .- (n .- numSignersB)) inA
+
+    -- Generate an second rewind that drops a suffix of A
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]:
```suggestion
    -- Generate a second rewind that drops a suffix of A
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 7]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
-     -- ^ (slot of latest signature, slot of next input)
-
--- | The output contains the specified number of 'InputSigner's and also some
--- additional 'InputEBB's
+genMockKey :: Int -> Gen (VerKeyDSIGN MockDSIGN)
+genMockKey numKeys = VerKeyMockDSIGN <$> choose (1, numKeys)
+
+-- | Generate a possibly empty sequence of 'InputEBB's and 'InputSigner's
+--
+-- INVARIANT The output contains exactly the specified number of 'InputSigner's.
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: `POSTCONDITION`?

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 8]
```diff
@@ -263,23 +362,31 @@ classifyWindow TestChainState{..}
...
+    Nothing            -> OriginLastA
+    Just InputEBB{}    -> EbbLastA
+    Just InputSigner{} -> SignedLastA
+
 {-------------------------------------------------------------------------------
   Tests
 -------------------------------------------------------------------------------}
 
--- Check that we are producing valid chain tests
+-- | Check that we are producing valid 'ChainStates'
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]:
```suggestion
-- | Check that we are producing valid 'PBftChainState's
```

File: ouroboros-consensus/test-util/Test/Util/Split.hs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  Right x -> Just x
+
+{-------------------------------------------------------------------------------
+  Properties
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "Test.Util.Split" $
+    [ testProperty "aux_spanLeft"    prop_spanLeft
+    , testProperty "aux_splitAtJust" prop_splitAtJust
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: Nice. Copy-paste error when I split these tests out of `PBFT.hs`.

File: ouroboros-consensus/test-util/Test/Util/Split.hs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+          Just (b, post) -> map Left pre ++ Right b : post
+
+    count = counterexample "wrong count" $
+            counterexample (show result) $
+            actual === expected
+      where
+        expected = length $ takeWhile isLeft xs
+        actual   = length pre
+
+    tick = counterexample "is not reverse" $
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: Renamed to `tickIsReversed`.

File: ouroboros-consensus/test-util/Test/Util/Split.hs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+--
+-- INVARIANT The output data is a segmentation of the given list.
+splitAtJust ::
+  forall x b.
+     HasCallStack
+  => (x -> Maybe b) -> Word64 -> [x] -> (Maybe ([x], b), [x])
+splitAtJust prj = \n xs ->
+  if 0 == n then (Nothing, xs)
+  else case peel xs of
+    (pre, Just (xb, xs')) -> Just `Arrow.first` go pre xb (n - 1) xs'
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: Done. Only wincing a little while adding the `bifunctor` to `build-depends` :p

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 12]
```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
...
-tests = testGroup "PBftChainState" [
-      testProperty "validGenerator"                   prop_validGenerator
-    , testProperty "appendPreservesInvariant"         prop_appendPreservesInvariant
-    , testProperty "rewindPreservesInvariant"         prop_rewindPreservesInvariant
-    , testProperty "rewindReappendId"                 prop_rewindReappendId
-    , testProperty "appendOldStatePreservesInvariant" prop_appendOldStatePreservesInvariant
-    , testProperty "appendOldStateRestoresPreWindow"  prop_appendOldStateRestoresPreWindow
+tests = testGroup "PBftChainState" $
+    -- these tests are cheap, so inflate the count
+    map (adjustOption $ \(QuickCheckTests n) -> QuickCheckTests $ 10 * n) $
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 6 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 8 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 7 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 13]
```diff
@@ -427,61 +572,87 @@ appendInputs ::
...
+  x : _ -> Just x
+
+safeLast :: [a] -> Maybe a
+safeLast
+ = go
+  where
+    go = \case
+      []     -> Nothing
+      [x]    -> Just x
+      _ : xs -> go xs
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs [hunk 14]
```diff
@@ -427,61 +572,87 @@ appendInputs ::
...
+  Just (InputSigner x) -> At $ CS.pbftSignerSlotNo x
 
 {-------------------------------------------------------------------------------
   Auxiliary
 -------------------------------------------------------------------------------}
 
+safeHead :: [a] -> Maybe a
+safeHead = \case
+  []    -> Nothing
+  x : _ -> Just x
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Yep; Edsko's comment said "edge cases", which made me think of off-by-ones.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Not in my opinion?

I changed it to

```
-- | The slot of the latest block
slotLatestInput :: LatestInput c -> WithOrigin SlotNo
slotLatestInput (LatestInput mbInp) = case mbInp of
  Nothing  -> Origin
  Just inp -> At $ slotInput inp
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: Fixed, thanks.

File: ouroboros-consensus/test-util/Test/Util/Split.hs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+  spanLeft
+-------------------------------------------------------------------------------}
+
+-- | The returned @b@ is the first in the list.
+--
+-- INVARIANT The output data is a segmentation of the given list.
+spanLeft ::
+  forall x a b.
+     HasCallStack
+  => (x -> Either a b) -> [x] -> ([a], Maybe (b, [x]))
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: I adopted your second suggestion for all of them in these two modules. Thanks.

**Comment by nfrisby** [comment]
I pushed up a commit that I think addresses all of your comments and Resolved the direct ones.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: This could inspire a QuickCheck feature request: a couple command-line parameters that

  * specify a total desired run-time

  * a minimum ratio from longest test to shortest test.

QC would inflate test counts for the quickest tests until the ratio was met. Then it would evenly increase all test counts (~batched round-robin?) until the desired total time expired.

File: ouroboros-consensus/test-util/Test/Util/Split.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: Isn't `Data.Bifunctor` part of `base`? See https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bifunctor.html.

File: ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs]: I would :+1: that feature request :smile:

**Comment by mrBliss** [comment]
After having a look at https://github.com/input-output-hk/ouroboros-network/pull/1337#discussion_r356479198 and squashing, you can bors this.

File: ouroboros-consensus/test-util/Test/Util/Split.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-util/Test/Util/Split.hs]: Yeeesh. Since 4.8 even. That's very good to know.

**Comment by nfrisby** [comment]
bors retry
